### ISSUE TITLE
Kubernetes CI Policy: release-blocking jobs must run in dedicated cluster

### ISSUE THREAD

**Issue by spiffxp** [kubernetes/test-infra#18549]
Part of https://github.com/kubernetes/test-infra/issues/18551

Why it's necessary:
- we believe declaring Guaranteed Pod QOS jobs may not be defended against Best Effort or Burstable Pods hogging all resources on the same node
- a cluster that only has Guaranteed pods is far more likely to respect resource requirements

What decisions do we need to make:
- For all of the jobs being migrated to k8s-infra-prow-build, we're going to use the same node pool as everything else.  To pin to a dedicated node pool will require much more boilerplate (or possibly augmenting prow's preset feature).  If after migrating everything we find we _do_ need a dedicated nodepool, then we'll pay the added cost.
- Jobs that can't be migrated quickly will remain in google.com-owned k8s-prow-builds until we can overcome whatever obstacles are keeping them there.  This means no community visibility into resource consumption, and test-infra-oncall will have to be relied upon.

Boskos projects to be added:
- [x] https://github.com/kubernetes/k8s.io/issues/1095 - gpu-project

Jobs to be migrated
- [x] https://github.com/kubernetes/test-infra/issues/19483 - ci-kubernetes-build (more involved due to https://github.com/kubernetes/k8s.io/issues/846; do this _after_ ci-kubernetes-build-fast)
- [x] https://github.com/kubernetes/test-infra/issues/19484 - ci-kubernetes-build-fast (more involved due to https://github.com/kubernetes/k8s.io/issues/846)
- [x] https://github.com/kubernetes/test-infra/pull/17674 - ci-kubernetes-conformance-kind-ga-only
- [x] https://github.com/kubernetes/test-infra/issues/18650 - ci-kubernetes-e2e-gce-device-plugin-gpu 
- [x] https://github.com/kubernetes/test-infra/pull/17488 - ci-kubernetes-e2e-gce-master-new-gci-kubectl-skew
- [x] https://github.com/kubernetes/test-infra/pull/17672 - ci-kubernetes-e2e-gci-gce
- [x] https://github.com/kubernetes/test-infra/pull/17488 - ci-kubernetes-e2e-gci-gce-alpha-features
- [x] https://github.com/kubernetes/test-infra/issues/18651 - ci-kubernetes-e2e-gci-gce-ingress
- [x] https://github.com/kubernetes/test-infra/pull/17488 - ci-kubernetes-e2e-gci-gce-reboot
- [x] https://github.com/kubernetes/test-infra/pull/17725 - ci-kubernetes-e2e-gci-gce-scalability
- [x] https://github.com/kubernetes/test-infra/pull/17887 - ci-kubernetes-e2e-ubuntu-gce-containerd
- [x] https://github.com/kubernetes/test-infra/pull/17488 - ci-kubernetes-gce-conformance-latest
- [x] https://github.com/kubernetes/test-infra/pull/17488 - ci-kubernetes-integration-master
- [x] https://github.com/kubernetes/test-infra/pull/17664 - ci-kubernetes-kind-e2e-parallel
- [x] https://github.com/kubernetes/test-infra/pull/17664 - ci-kubernetes-kind-ipv6-e2e-parallel
- [x] https://github.com/kubernetes/test-infra/pull/17488 - ci-kubernetes-node-kubelet
- [x] https://github.com/kubernetes/test-infra/pull/17488 - ci-kubernetes-verify-master
- [x] https://github.com/kubernetes/test-infra/pull/20964 - ~periodic-kubernetes-bazel-build-master~ (was moved to release-infoming)
- [x] https://github.com/kubernetes/test-infra/issues/18652 - periodic-kubernetes-bazel-test-master

TODO:
- [x] https://github.com/kubernetes/test-infra/pull/18576 - write a test that enforces this policy, but logs instead of errors
- [x] determine which jobs may not be able to be migrated quickly, decide if / how to mitigate
- [x] migrate all jobs that can be migrated quickly
- [x] migrate jobs blocked on https://github.com/kubernetes/k8s.io/issues/846
- [x] migrate jobs that need to be migrated away from RBE
- [x] https://github.com/kubernetes/test-infra/pull/20989 - make policy enforcement test blocking (fails instead of logs)
- [x] declare we are finished (or done enough to move on)

**Comment by BenTheElder** [comment]
NodePool is probably fine, though it has the downside that you have to explicitly set a matching node selector and taint toleration in every prowjob podspec if you want to actually make it dedicated. there's no abstraction for this, and they're a little verbose.

we've done this before when experimenting with ubuntu nodes for kind IPv6 before we'd standardized on that.

**Comment by spiffxp** [comment]
Current status (not counting issues that are held open to monitor jobs)
```
spiffxp@spiffxp-macbookpro:test-infra (master %)$ go test -v -count=1 ./config/tests/jobs
# ...
=== RUN   TestKubernetesReleaseBlockingJobsShouldRunOnK8sInfraProwBuild
    TestKubernetesReleaseBlockingJobsShouldRunOnK8sInfraProwBuild: jobs_test.go:1084: ci-kubernetes-build: should run on cluster: k8s-infra-prow-build, found: default
    TestKubernetesReleaseBlockingJobsShouldRunOnK8sInfraProwBuild: jobs_test.go:1084: ci-kubernetes-build-1-19: should run on cluster: k8s-infra-prow-build, found: default
    TestKubernetesReleaseBlockingJobsShouldRunOnK8sInfraProwBuild: jobs_test.go:1084: ci-kubernetes-build-fast: should run on cluster: k8s-infra-prow-build, found: default
    TestKubernetesReleaseBlockingJobsShouldRunOnK8sInfraProwBuild: jobs_test.go:1084: ci-kubernetes-build-stable1: should run on cluster: k8s-infra-prow-build, found: default
    TestKubernetesReleaseBlockingJobsShouldRunOnK8sInfraProwBuild: jobs_test.go:1084: ci-kubernetes-build-stable2: should run on cluster: k8s-infra-prow-build, found: default
    TestKubernetesReleaseBlockingJobsShouldRunOnK8sInfraProwBuild: jobs_test.go:1084: ci-kubernetes-build-stable3: should run on cluster: k8s-infra-prow-build, found: default
    TestKubernetesReleaseBlockingJobsShouldRunOnK8sInfraProwBuild: jobs_test.go:1084: periodic-kubernetes-bazel-build-1-16: should run on cluster: k8s-infra-prow-build, found: default
    TestKubernetesReleaseBlockingJobsShouldRunOnK8sInfraProwBuild: jobs_test.go:1084: periodic-kubernetes-bazel-build-1-17: should run on cluster: k8s-infra-prow-build, found: default
    TestKubernetesReleaseBlockingJobsShouldRunOnK8sInfraProwBuild: jobs_test.go:1084: periodic-kubernetes-bazel-build-1-18: should run on cluster: k8s-infra-prow-build, found: default
    TestKubernetesReleaseBlockingJobsShouldRunOnK8sInfraProwBuild: jobs_test.go:1084: periodic-kubernetes-bazel-build-1-19: should run on cluster: k8s-infra-prow-build, found: default
    TestKubernetesReleaseBlockingJobsShouldRunOnK8sInfraProwBuild: jobs_test.go:1084: periodic-kubernetes-bazel-build-master: should run on cluster: k8s-infra-prow-build, found: default
    TestKubernetesReleaseBlockingJobsShouldRunOnK8sInfraProwBuild: jobs_test.go:1084: periodic-kubernetes-bazel-test-1-16: should run on cluster: k8s-infra-prow-build, found: default
    TestKubernetesReleaseBlockingJobsShouldRunOnK8sInfraProwBuild: jobs_test.go:1084: periodic-kubernetes-bazel-test-1-17: should run on cluster: k8s-infra-prow-build, found: default
    TestKubernetesReleaseBlockingJobsShouldRunOnK8sInfraProwBuild: jobs_test.go:1084: periodic-kubernetes-bazel-test-1-18: should run on cluster: k8s-infra-prow-build, found: default
    TestKubernetesReleaseBlockingJobsShouldRunOnK8sInfraProwBuild: jobs_test.go:1084: periodic-kubernetes-bazel-test-1-19: should run on cluster: k8s-infra-prow-build, found: default
--- PASS: TestKubernetesReleaseBlockingJobsShouldRunOnK8sInfraProwBuild (0.02s)
```

Will need to decide how we'd like to define exceptions for the build jobs, and/or dedicate resources to them.

**Comment by justaugustus** [comment]
Assigning Dan to TL completion of this.
/assign @hasheddan

**Comment by spiffxp** [comment]
Updated description to enumerate all release-blocking jobs and the umbrella issues tracking their progress or the PRs that migrated them

**Comment by spiffxp** [comment]
Current status...
```
spiffxp@spiffxp-macbookpro:test-infra (master %)$ go test -v -count=1 ./config/tests/jobs/...
# ...
=== RUN   TestKubernetesReleaseBlockingJobsShouldRunOnK8sInfraProwBuild
    jobs_test.go:1049: ci-kubernetes-build: should run on cluster: k8s-infra-prow-build, found: default
    jobs_test.go:1049: ci-kubernetes-build-1-19: should run on cluster: k8s-infra-prow-build, found: default
    jobs_test.go:1049: ci-kubernetes-build-1-20: should run on cluster: k8s-infra-prow-build, found: default
    jobs_test.go:1049: ci-kubernetes-build-stable2: should run on cluster: k8s-infra-prow-build, found: default
    jobs_test.go:1049: ci-kubernetes-build-stable3: should run on cluster: k8s-infra-prow-build, found: default
    jobs_test.go:1049: periodic-kubernetes-bazel-build-1-17: should run on cluster: k8s-infra-prow-build, found: default
    jobs_test.go:1049: periodic-kubernetes-bazel-build-1-18: should run on cluster: k8s-infra-prow-build, found: default
    jobs_test.go:1049: periodic-kubernetes-bazel-build-1-19: should run on cluster: k8s-infra-prow-build, found: default
    jobs_test.go:1049: periodic-kubernetes-bazel-build-1-20: should run on cluster: k8s-infra-prow-build, found: default
    jobs_test.go:1049: periodic-kubernetes-bazel-build-master: should run on cluster: k8s-infra-prow-build, found: default
```

We have "canary" versions of these jobs that run on k8s-infra-prow-build, and write to `gs://k8s-release-dev`.

However, there is much out there that depends on the existing jobs that write to `gs://kubernetes-release-dev` and `gcr.io/kubernetes-ci-images` (ref: https://github.com/kubernetes/test-infra/issues/19483#issuecomment-733277378).  I'm considering a KEP for deprecation/migration/removal.

I think full completion of the deprecation process is out of scope for this issue.  So I would like us to identify what the minimum set of jobs/scripts/etc should be pulling from `gs://k8s-release-dev` to call this done.

**Comment by spiffxp** [comment]
I think the following PRs can close this out:
- https://github.com/kubernetes/test-infra/pull/20885 - enforced that all release-blocking jobs pull from gs://k8s-release-dev
- https://github.com/kubernetes/test-infra/pull/20963 - proposes moving bazel-build jobs to release-informing
- https://github.com/kubernetes/test-infra/pull/20964 - proposes renaming build jobs to build-deprecated and moving to release-informing, renaming build-canary jobs to build and moving to release-blocking
- TBD - a final PR to move "should" to "must" for `TestKubernetesReleaseBlockingJobsShouldRunOnK8sInfraProwBuild` (rename, fail instead of log)

**Comment by spiffxp** [comment]
/assign
/assign @justaugustus 
Can you think of anything else that needs to happen to call this done?

**Comment by spiffxp** [comment]
Tangentially related: followup work to deprecate google-hosted artifacts related to release-blocking builds will be tracked under https://github.com/kubernetes/k8s.io/issues/1571

**Comment by spiffxp** [comment]
/close
Calling this done!
